pkg/providers/imagefamily/image.go

```diff
@@ -91,57 +86,116 @@ func (p *Provider) KubeServerVersion(ctx context.Context) (string, error) {
 	return version, nil
 }
 
-// Input versionName == "" to get the latest version
-func (p *Provider) GetImageID(ctx context.Context, communityImageName, publicGalleryURL, versionName string) (string, error) {
-	key := fmt.Sprintf("%s/%s/%s", publicGalleryURL, communityImageName, versionName)
-	imageID, found := p.imageCache.Get(key)
-	if found {
-		return imageID.(string), nil
-	}
-
-	if versionName == "" {
-		pager := p.imageVersionsClient.NewListPager(p.location, publicGalleryURL, communityImageName, nil)
-		topImageVersionCandidate := armcompute.CommunityGalleryImageVersion{}
-		for pager.More() {
-			page, err := pager.NextPage(context.Background())
-			if err != nil {
-				return "", err
-			}
-			for _, imageVersion := range page.CommunityGalleryImageVersionList.Value {
-				if lo.IsEmpty(topImageVersionCandidate) || imageVersion.Properties.PublishedDate.After(*topImageVersionCandidate.Properties.PublishedDate) {
-					topImageVersionCandidate = *imageVersion
-				}
-			}
+// Get returns Image ID for the given instance type. Images may vary due to architecture, accelerator, etc
+func (p *Provider) Get(ctx context.Context, nodeClass *v1alpha2.AKSNodeClass, instanceType *cloudprovider.InstanceType, imageFamily ImageFamily) (string, error) {
+	defaultImages := imageFamily.DefaultImages()
+	for _, defaultImage := range defaultImages {
+		if err := instanceType.Requirements.Compatible(defaultImage.Requirements, v1alpha2.AllowUndefinedLabels); err == nil {
+			return p.GetLatestImageID(ctx, defaultImage, nodeClass.Spec.GetImageVersion())
 		}
-		versionName = lo.FromPtr(topImageVersionCandidate.Name)
 	}
 
-	selectedImageID := BuildImageID(publicGalleryURL, communityImageName, versionName)
-	if p.cm.HasChanged(key, selectedImageID) {
-		logging.FromContext(ctx).With("image-id", selectedImageID).Info("discovered new image id")
+	return "", fmt.Errorf("no compatible images found for instance type %s", instanceType.Name)
+}
+
+// GetLatestImageID returns the latest image ID for the given image version and image specification
+func (p *Provider) GetLatestImageID(ctx context.Context, defaultImage DefaultImageOutput, imageVersion string) (string, error) {
+	// Managed Karpenter will use the AKS Managed Shared Image Galleries
+	if options.FromContext(ctx).ManagedKarpenter {
```

Comment:
Can one cache here, instead of later?

Reply:
I don't get what you mean.

Reply:
Right now caching is implemented within individual `getImageIDSIG` and `getImageIDCIG`. What I am wondering is whether caching can instead be pull up here, into GetLatestImageID, simplifying the individual function logic.

 (Separately, wondering if `getSIGImageID` and `getCIGImageID` names would make it easier to distinguish the two?)

Reply:
Refactored the code here: https://github.com/Azure/karpenter-provider-azure/pull/526/commits/8c5fd4a1c6cb5486c820691c7eebb980a24828ba